detox/src/devices/drivers/android/emulator/EmulatorDeviceAllocator.js

```diff
@@ -0,0 +1,44 @@
+const AndroidDeviceAllocator = require('../AndroidDeviceAllocator');
+const logger = require('../../../../utils/logger').child({ __filename });
+
+const DetoxEmulatorsPortRange = {
+  min: 10000,
+  max: 20000
+};
+
+// TODO unit test
+class EmulatorDeviceAllocator extends AndroidDeviceAllocator {
+  constructor(deviceRegistry, freeDeviceFinder) {
+    super(deviceRegistry, logger);
+    this._freeDeviceFinder = freeDeviceFinder;
+  }
+
+  async _doAllocateDevice(deviceQuery) {
+    let placeholderPort = null;
+    let adbName = null;
+
+    await this.deviceRegistry.allocateDevice(async () => {
+      adbName = await this._freeDeviceFinder.findFreeDevice(deviceQuery);
+      if (!adbName) {
+        placeholderPort = this._allocateEmulatorPlaceholderPort();
+        adbName = `emulator-${placeholderPort}`;
+      }
+      return adbName;
```

Comment:
DeviceRegistry.allocateDevice could use a sort of `.toString()` method to extract the unique id, so you don't have to run these perversions with scoped `let` variables.